client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js

```diff
@@ -1,129 +1,429 @@
+} from 'components/Register';
 import Title from '../../Register/Title';
+import Config from 'lib/Config'; // eslint-disable-line
+import confirm from '@silverstripe/reactstrap-confirm';
 
 const fallbacks = require('../../../../lang/src/en.json');
 
 class RegisteredMFAMethodListField extends Component {
   constructor(props) {
     super(props);
 
+    const { initialAvailableMethods, registeredMethods } = props;
+
+    // Move registered methods into state as we might remove and add them during the lifetime of
+    // this component.
     this.state = {
       modalOpen: false,
+      registeredMethods,
     };
 
-    props.onUpdateAvailableMethods(props.availableMethods);
+    props.onUpdateAvailableMethods(initialAvailableMethods);
 
     this.handleToggleModal = this.handleToggleModal.bind(this);
+    this.handleClickRemove = this.handleClickRemove.bind(this);
+    this.handleAddMethod = this.handleAddMethod.bind(this);
+    this.handleRemoveBackupMethod = this.handleRemoveBackupMethod.bind(this);
+    this.handleRemoveMethod = this.handleRemoveMethod.bind(this);
+    this.handleResetMethod = this.handleResetMethod.bind(this);
   }
 
-  getAddMethodButtonLabel() {
-    const { ss: { i18n } } = window;
-    const { registeredMethods } = this.props;
+  componentDidUpdate() {
+    const { registrationScreen } = this.props;
+    const { modalOpen } = this.state;
 
-    return registeredMethods.length
-      ? i18n._t(
-        'MultiFactorAuthentication.ADD_ANOTHER_METHOD',
-        fallbacks['MultiFactorAuthentication.ADD_ANOTHER_METHOD']
-      )
-      : i18n._t(
-        'MultiFactorAuthentication.ADD_FIRST_METHOD',
-        fallbacks['MultiFactorAuthentication.ADD_FIRST_METHOD']
-      );
+    // Close the modal if the user returns to the introduction screen (ie. they click "back").
+    if (registrationScreen === SCREEN_INTRODUCTION && modalOpen) {
+      this.handleToggleModal();
+    }
+  }
+
+  /**
+   * The backup and default methods are rendered separately
+   * @returns {Array<object>}
+   */
+  getBaseMethods() {
+    const { backupMethod, defaultMethod } = this.props;
+    let { registeredMethods: methods } = this.state;
+
+    if (backupMethod) {
+      methods = methods.filter(method => method.urlSegment !== backupMethod.urlSegment);
+    }
+
+    if (defaultMethod) {
+      methods = methods.filter(method => method.urlSegment !== defaultMethod.urlSegment);
+    }
+
+    return methods;
   }
 
   /**
-   * Handles an event/request to toggle the visibility of the register modal.
+   * Handle a request to toggle the modal. This method will also reset the modal provided it's being
+   * opened, unless `true` is provided as the parameter
+   *
+   * @param {boolean} skipReset
    */
-  handleToggleModal() {
+  handleToggleModal(skipReset = false) {
     const { modalOpen } = this.state;
 
     this.setState({
       modalOpen: !modalOpen,
     });
 
-    if (!modalOpen) {
+    if (!modalOpen && skipReset !== true) {
```

Comment:
Can we just say `!skipReset`? It seems like you're implementing your own kind of strict typing logic here, but not sure it's really that necessary

Reply:
Yeah I don't know why I did that.